Docs: Update GitHub SSO

alexfornuto opened the pull request:
Closes #21986 

Updates include:

- Use `tctl sso configure` to generate connector resource.
   - **Note:** Instructions should be updated when/if #26786 is addressed.
-  Update `cap` instructions for current default existence of the resource, and edit to be optional (you can specify the auth type otherwise), and to use `tctl edit`.
- Updated screenshot for default light theme
- **CLI REFERENCE** - Fixed incorrect description of `--teams-to-roles` flag.

Commit:
Fix erroneous command reference

Commit:
update github SSO doc

Commit:
include tctl sso test

zmb3 commented on docs/pages/access-controls/sso/github-sso.mdx:
```diff
@@ -32,86 +32,110 @@ https://PROXY_ADDRESS/v1/webapi/github/
-Define a GitHub authentication connector by creating a file called `github.yaml`
-with the following content:
+Define a GitHub authentication connector using `tctl`. Update this example
+command with your Oauth app's client ID, client secret, and the roles you want
```
```suggestion
command with your OAuth app's client ID, client secret, and the roles you want
```

zmb3 commented on docs/pages/access-controls/sso/github-sso.mdx:
````diff
@@ -32,86 +32,110 @@ https://PROXY_ADDRESS/v1/webapi/github/
-<Tabs>
-<TabItem label="Open Source" scope="oss">
+```code
+$ tctl ssh configure github \
````
```suggestion
$ tctl sso configure github \
```

zmb3 commented on docs/pages/access-controls/sso/github-sso.mdx:
````diff
@@ -253,8 +236,60 @@ You will receive the details of your user session within the CLI.
+Edit the existing `cluster_auth_preference` resource using `tctl`:
+
+```code
+$ tctl edit cap
````
Note: `tctl edit` is not available in v11, so you'll have to change this section for the v11 backport.

Tener replied:
I believe what `tctl edit` does is something along these lines:

1. `tctl get cap > cap.yaml`
2. `$EDITOR cap.yaml`
3. `tctl create -f cap.yaml`

zmb3 commented on docs/pages/reference/cli.mdx:
````diff
@@ -2078,7 +2078,7 @@ Configure the GitHub auth connector.
+In this example, members of the `devs` team in the GitHub org `octocats` will be assigned the `access`, and `editor` roles in Teleport.
 
 ```code
 $ tctl sso test --id=GITHUB_CLIENT_ID --secret=GITHUB_SECRET --teams-to-roles=octocats,devs,access,editor [<other-flags>]
````
I don't see a `--teams-to-roles` flag on the `tctl sso test` command. Something seems off here.

alexfornuto replied:
Yea, this should probably be `tctl ssh configure github`. CC @Tener

Tener replied:
Yep, this should be `tctl sso configure github` as per section header.

Commit:
Apply suggestions from code review

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

Commit:
Apply suggestions from code review

Tener commented on docs/pages/access-controls/sso/github-sso.mdx:
```diff
@@ -32,86 +32,110 @@ https://PROXY_ADDRESS/v1/webapi/github/
+</Details>
 
-Teleport will request only the `read:org` OAuth scope. Read more about OAuth scopes in GitHub's documentation:
+You can optionally test the connector configuration before applying it to your cluster:
```
I'd strongly encourage the testing to be done prior to applying the changes.

alexfornuto replied:
resolved in 8c3746791ea7a82a6a6cbba16d68869762f7a451

Tener commented on docs/pages/access-controls/sso/github-sso.mdx:
````diff
@@ -32,86 +32,110 @@ https://PROXY_ADDRESS/v1/webapi/github/
 ```code
 $ tctl create github.yaml
-# authentication connector "github" has been created
+authentication connector "github" has been created
````
On the second attempt, this will fail as the resource will already exist by that point. I wonder if we should mention the `-f` option here to solve that?

alexfornuto replied:
Adding `-f`, which is standard (I forgot) in docs for this very reason. Regarding alternate flows, I think at some point we have to trust the reader to know this sort of bash piping syntax and use whatever they're comfortable with.

alexfornuto replied:
Resolved in 9dafbcfabff402479db7a487e4f58a52ea44b446

Tener commented on docs/pages/access-controls/sso/github-sso.mdx:
```diff
@@ -253,8 +236,60 @@ You will receive the details of your user session within the CLI.
+## Step 4/4. Configure authentication preference
+
+In the previous step we signed in to Teleport using GitHub credentials by
+specifying GitHub as our auth type. By editing the `cluster_auth_preference`
+resource.
```
The sentence isn't finished; something like this?

```suggestion
specifying GitHub as our auth type. By editing the `cluster_auth_preference`
resource we can make it the default auth type.
```

Tener commented on docs/pages/access-controls/sso/github-sso.mdx:
```diff
@@ -32,86 +32,110 @@ https://PROXY_ADDRESS/v1/webapi/github/
-[Creating an OAuth App](https://docs.github.com/en/developers/apps/building-oauth-apps/creating-an-oauth-app)
+Create a client secret to use along with the client ID in the next step.
 
 ## Step 2/3. Create a GitHub authentication connector
 
-Define a GitHub authentication connector by creating a file called `github.yaml`
-with the following content:
+Define a GitHub authentication connector using `tctl`. Update this example
+command with your OAuth app's client ID, client secret, and the roles you want
+to map from your GitHub organization to Teleport roles. See [tctl sso configure
+github](../../reference/cli.mdx#tctl-sso-configure-github) for a full reference
+of flags for this command:
```
We say that we need client id, client secret, org name and team name, but if someone isn't well versed in the lingo, they will be lost. Perhaps we could present a short table explaining the necessary tokens, what they mean, and where they come from.

alexfornuto replied:
Resolved in 0c4879515c00988313f04af6bd9795e81b99d987

Tener commented on docs/pages/access-controls/sso/github-sso.mdx:
```diff
@@ -32,86 +32,110 @@ https://PROXY_ADDRESS/v1/webapi/github/
-      team: admins # GitHub team name within that organization
-      # Maps octocats/admins to the "access" Teleport role
+    - organization: ORG-NAME
       roles:
         - access
+        - editor
+      team: GITHUB-TEAM
```
It isn't immediately clear how can you utilize multiple team-to-role mappings, which will be a common use case. Do you mind adding a `<Details>` section covering that?

I think the easiest way to do so is via multiple instances of `--teams-to-roles` flag, like so:

```
> tctl sso configure github --id GITHUB_CLIENT_ID --secret GITHUB_SECRET -r octocats,devs,access,auditor,editor -r foo,bar,access

kind: github
metadata:
  name: github
spec:
  api_endpoint_url: ""
  client_id: GITHUB_CLIENT_ID
  client_secret: GITHUB_SECRET
  display: ""
  endpoint_url: ""
  redirect_url: https://boson.tener.io:3080/v1/webapi/github/callback
  teams_to_logins: null
  teams_to_roles:
  - organization: octocats
    roles:
    - access
    - auditor
    - editor
    team: devs
  - organization: foo
    roles:
    - access
    team: bar
version: v3
```

If someone wants to edit the YAML file manually, they can also pick up the syntax from the example above.

alexfornuto replied:
Resolved in b2cd72ce7c2dc70e1c2df62107ef396d1a6d26a7

Commit:
clarify source for connectory vars...

Responding to this feedback: https://github.com/gravitational/teleport/pull/26790/files\#r1206440680

Commit:
document multiple role mappings

in response to feedback: https://github.com/gravitational/teleport/pull/26790/files\#r1206446669

Commit:
Strongly encourage testing

responding to https://github.com/gravitational/teleport/pull/26790/files\#r1206421549

Commit:
use tctl with -f

Commit:
Update docs/pages/access-controls/sso/github-sso.mdx

Co-authored-by: Krzysztof Skrzętnicki <krzysztof.skrzetnicki@goteleport.com>

Tener reviewed (approved):
👍

alexfornuto commented on docs/pages/access-controls/sso/github-sso.mdx:
```diff
@@ -136,63 +201,17 @@ Here is an example role configuration snippet using the trait variable:
-spec:
-  type: github
-  webauthn:
-    rp_id: 'example.teleport.sh'
```
@Tener while reviewing merge conflicts with `branch/v12` I realized that the configuration created by `tctl sso configure github` does not include `spec.webauthn.rp_id`. Is this something that the `tctl `command can include but is undocumented, and/or should we instruct the reader to manually add to `cap`?
